doctor/lib/text_extraction.py

```diff
@@ -0,0 +1,326 @@
+        output_type=Output.DICT,
+    )
+    df = pd.DataFrame(data_dict)
+    filtered_data = df[(df.conf != -1)]
+    block_ids = (
+        filtered_data.groupby("block_num")
+        .first()
+        .sort_values("top")
+        .index.tolist()
+    )
+    blocks = [
+        filtered_data[filtered_data["block_num"] == block]
+        for block in block_ids
+    ]
+    return blocks
+
+
+def extract_with_ocr(page: pdfplumber.pdf.Page, strip_margin: bool) -> str:
+    """Extract the page using OCR
+
+    :param page:Pdf Plumber Page
+    :param strip_margin: If we should trim the margins
+    :return: The extracted content for the page
+    """
+
+    image = convert_pdf_page_to_image(page, strip_margin)
+    data = ocr_image_to_data(image)
+    content = ""
+    prev = {}
+    for words in data:
+        for index, word in words.iterrows():
+            content = insert_whitespace(content, word, prev)
+            content += get_word(word, image.size[0], strip_margin)
+            prev = word
+    content = cleanup_content(content, page.page_number)
+    return content
+
+
+def insert_whitespace(content: str, word: dict, prev: dict) -> str:
+    """Insert whitespace after or before word
+
+    :param content: The text extracted so far
+    :param word: The OCR extraction object
+    :param prev: The previous word object extracted
+    :return: The content with the whitespace appended
+    """
+    is_new_line = prev.get("line_num", 0) != word["line_num"]
+    is_new_par = prev.get("par_num", 0) != word["par_num"]
+    prev_end = prev.get("left", 1) + prev.get("width", 1)
+
+    # Add vertical whitespace
+    if is_new_line or is_new_par:
+        vertical_gap = word["top"] - (
+            prev.get("top", 0) + prev.get("height", 0)
+        )
+        content += "\n\n" if vertical_gap > 100 else "\n"
+        prev_end = 0
+
+    # add horizontal whitespace
+    content += " " * int(((word["left"] - prev_end) / 25))
+    return content
+
+
+def get_word(word_dict: dict, width: float, strip_margin: bool) -> str:
+    """Append word to content
+
+    :param word_dict: the word object from tesseract
+    :param width: The width of the document
+    :param strip_margin: should we strip the margin
+    :return: The text with space
+    """
+    if strip_margin:
+        left_margin = (1 / 8.5) * width
+        right_margin = (7.5 / 8.5) * width
+    else:
+        left_margin = (0.5 / 8.5) * width
+        right_margin = (8.0 / 8.5) * width
+
+    word = word_dict["text"]
+    conf = word_dict["conf"]
+    if word_dict["left"] + word_dict["width"] < left_margin and conf < 40:
+        word = " " * len(word)
+    elif (conf == 0 and len(word) < 4) or word_dict["left"] == 0:
+        word = " " * len(word)
+    elif conf < 5 and (len(word) < 4 or len(word) > 20):
+        word = "□" * len(word)
+    elif conf < 40 and word_dict["left"] > right_margin:
+        word = "□" * len(word)
```

Comment:
We really need some explanation of what's going on here or else it's entirely unclear. LOTS of magic numbers to explain.

Reply:
sounds good.